scipy/spatial/ckdtree/src/build.cxx

```diff
@@ -12,29 +12,76 @@
 #include <typeinfo>
 #include <stdexcept>
 #include <ios>
+#include <future>
+#include <atomic>
+#include <list>
+#include <memory>
 
 #define tree_buffer_root(buf) (&(buf)[0][0])
 
+#define STD_VECTOR_RESERVE
+
+static ckdtree_intp_t
+build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+      double *maxes, double *mins,
+      const int _median, const int _compact,
+      std::list<std::vector<ckdtreenode>>* buflist, int workers);
+
+struct build_worker {
+    std::list<std::vector<ckdtreenode>> buflist_;
+    std::vector<double> mins_, maxes_;
+    std::future<ckdtree_intp_t> future_;
+
+    ckdtree_intp_t run(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+          double *maxes, double *mins,
+          const int _median, const int _compact,
+          int workers)
+    {
+        const int subworkers = workers/2;
+        if(subworkers > 0)
+        {
+            mins_.resize(self->m);
+            std::copy_n(mins, mins_.size(), mins_.begin());
+            maxes_.resize(self->m);
+            std::copy_n(maxes, maxes_.size(), maxes_.begin());
+
+            future_ = std::async(std::launch::async, [&]() {
+                    buflist_.emplace_back();
+#ifdef STD_VECTOR_RESERVE
+                    buflist_.back().reserve((end_idx-start_idx)/self->leafsize/subworkers);
+#endif
```

Comment:
I am not actually suggesting to leave this conditional compilation in: Should we `.reserve()` or not? @sturlamolden  @peterbell10

Reply:
My vote is for reserve.

I think the worst case scenario is an unbalanced tree where half the threads threads immediately terminate with a leaf node and the other half holds almost all of the inner nodes. Even in this worst-case, I think the memory usage would only be ~1.5x that of a balanced tree. I don't think that's too unreasonable. Especially as it's only a temporary spike in allocated memory.

Reply:
Removed the guards and left the reserve.